File: worker/src/feeds/polymarket.ts
/** Polymarket Gamma API client (keyless, browser UA required). Mirrors bot/feeds/polymarket.py. */
import type { Http, LiveEvent, MarketPrice, OutcomePrice, Source } from "./base.ts";

const GAMMA = "https://gamma-api.polymarket.com";

interface GammaMarket {
  outcomes?: string | string[];
  outcomePrices?: string | string[];
  volume?: string | number;
}
interface GammaEvent {
  id?: string | number;
  title?: string;
  markets?: GammaMarket[];
}

function asArray(v: string | string[] | undefined): string[] {
  if (!v) return [];
  if (Array.isArray(v)) return v;
  try {
    const parsed = JSON.parse(v) as unknown;
    return Array.isArray(parsed) ? parsed.map(String) : [];
  } catch {
    return [];
  }
}

export class PolymarketSource implements Source {
  readonly name = "polymarket";

  async fetchLive(sport: string, http: Http): Promise<LiveEvent[]> {
    // No keyword prefilter: matching to fixtures happens in monitor.ts
    // via team-token overlap (same as the Python bot).
    let data: GammaEvent[];
    try {
      data = await http.getJson<GammaEvent[]>(`${GAMMA}/events`, { closed: "false", limit: "150" });
    } catch {
      return [];
    }
    const events: LiveEvent[] = [];
    for (const ev of data ?? []) {
      const markets: MarketPrice[] = [];
      for (const m of ev.markets ?? []) {
        try {
          const outcomes = asArray(m.outcomes);
          const prices = asArray(m.outcomePrices);
          if (outcomes.length === 0 || prices.length === 0) continue;
          const volume = Number(m.volume ?? 0) || 0;
          const ops: OutcomePrice[] = [];
          for (let i = 0; i < Math.min(outcomes.length, prices.length); i++) {
            const p = Number(prices[i]);
            if (!(p > 0)) continue;
            ops.push({ name: String(outcomes[i]), decimalOdds: 1 / p, volume });
          }
          if (ops.length >= 2) markets.push({ marketType: "moneyline", line: null, outcomes: ops });
        } catch {
          continue;
        }
      }
      if (markets.length > 0) {
        events.push({
          eventId: `poly-${ev.id}`,
          sport,
          matchLabel: ev.title ?? "?",
          isLive: true,
          markets,
          rawScore: "",
        });
      }
    }
    return events;
  }
}

File: bot/feeds/polymarket.py
"""Polymarket Gamma API client (keyless, browser UA required)."""
from __future__ import annotations

from .base import LiveEvent, MarketPrice, OutcomePrice, Source
from .http import get_json

GAMMA = "https://gamma-api.polymarket.com"

# sport key -> keyword filter used against event titles/series
SPORT_KEYWORDS = {
    "soccer-epl": ["premier league", "epl"],
    "soccer-laliga": ["la liga", "laliga"],
    "soccer-bundesliga": ["bundesliga"],
    "soccer-seriea": ["serie a"],
    "soccer-ligue1": ["ligue 1"],
    "soccer-ucl": ["champions league"],
    "nba": ["nba"],
    "nfl": ["nfl"],
    "mlb": ["mlb"],
    "nhl": ["nhl"],
    "tennis-atp": ["atp", "tennis"],
}


class PolymarketSource(Source):
    name = "polymarket"

    def fetch_live(self, sport: str) -> list[LiveEvent]:
        # No keyword prefilter: sports liquidity moves across series/titles.
        # Matching to fixtures happens in monitor.py via team-token overlap.
        try:
            data = get_json(f"{GAMMA}/events",
                            params={"closed": "false", "limit": 150})
        except Exception:
            return []
        events: list[LiveEvent] = []
        for ev in data or []:
            title = ev.get("title") or ""
            markets: list[MarketPrice] = []
            for m in ev.get("markets") or []:
                try:
                    outcomes = m.get("outcomes")
                    prices = m.get("outcomePrices")
                    if isinstance(outcomes, str):
                        import json as _json
                        outcomes = _json.loads(outcomes)
                        prices = _json.loads(prices)
                    if not outcomes or not prices:
                        continue
                    volume = float(m.get("volume") or 0)
                    ops = [OutcomePrice(name=str(o),
                                        decimal_odds=(1 / float(p)) if float(p) > 0 else 0.0,
                                        volume=volume)
                           for o, p in zip(outcomes, prices) if float(p) > 0]
                    if len(ops) >= 2:
                        markets.append(MarketPrice(market_type="moneyline",
                                                   line=None, outcomes=ops))
                except Exception:
                    continue
            if markets:
                events.append(LiveEvent(
                    event_id=f"poly-{ev.get('id')}",
                    sport=sport,
                    match_label=ev.get("title", "?"),
                    is_live=True,
                    markets=markets,
                ))
        return events
